src/lib/datepicker/datepicker-input.ts

```diff
@@ -319,17 +319,18 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
     }
   }
 
-  _onInput(value: string) {
+  _onInput(value: string): void {
     let date = this._dateAdapter.parse(value, this._dateFormats.parse.dateInput);
     this._lastValueValid = !date || this._dateAdapter.isValid(date);
     date = this._getValidDateOrNull(date);
 
     if (!this._dateAdapter.sameDate(date, this._value)) {
       this._value = date;
-      this._cvaOnChange(date);
       this._valueChange.emit(date);
       this.dateInput.emit(new MatDatepickerInputEvent(this, this._elementRef.nativeElement));
     }
+
+    this._cvaOnChange(date);
```

Comment:
We probably want to call `_cvaOnChange` before emitting. How about doing it like this:

```ts
this._value = date;
this._cvaOnChange(date);

if (!this._dateAdapter.sameDate(date, this._value)) {
  this._valueChange.emit(date);
  this.dateInput.emit(new MatDatepickerInputEvent(this, this._elementRef.nativeElement));
}
```

Reply:
Sounds good

Reply:
Actually, this won't work. You will end up never calling the dateInput output since this._value will always be value.

Reply:
Oh, yeah you can just move the check before

```ts
const shouldEmit = this._dateAdapter.sameDate(date, this._valie);
```